From 1028497550fd7c45147e9c5e410c5f5da14721cb Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Mon, 18 Dec 2023 17:50:32 +0200
Subject: [PATCH] [SW-166373] gaudi2: fix tpm boot failure

The TPM is held at reset until the deassertion of
prstn signal, hence accessing tpm before the
deassertion causes the transaction failure. This
patch adds support to wait for deassertion of
prstn signal for 200ms before accessing TPM.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I331a614960e5e0834f3e37aad864976a1d6ab78d
---

diff --git a/arch/gaudi2/common/tpm_auth.c b/arch/gaudi2/common/tpm_auth.c
index 0adfc82..27021c2 100644
--- a/arch/gaudi2/common/tpm_auth.c
+++ b/arch/gaudi2/common/tpm_auth.c
@@ -13,6 +13,9 @@
 #include <boot_desc.h>
 #include <errno.h>
 #include "lzo/unaligned.h"
+#include <pll.h>
+#include <gaudi2_fw_if.h>
+#include <psoc_global_conf_regs.h>
 
 #define TPM_NV_DEV_SN_ADDRESS		0x01c10100
 
@@ -38,9 +41,20 @@
 	return 0;
 }
 
+#define TPM_PRSTN_TIMEOUT_US	200000	/* 200msec */
 int tpm_init(bool print)
 {
 	int ret;
+	u64 start = timestamp_get_us();
+
+	/* TPM is held in reset until PRSTN is de-asserted */
+	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN)) {
+		if (timeout_expired(start, TPM_PRSTN_TIMEOUT_US)) {
+			preboot_puts("TPM: Timeout waiting for PRSTN\n");
+			cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+			return -ETIMEDOUT;
+		}
+	}
 
 	ret = tpm2_init(print);
 	if (ret) {
